homeassistant/components/growatt_server/coordinator.py

```diff
@@ -0,0 +1,215 @@
+            elif self.device_type == "tlx":
+                tlx_info = await self.hass.async_add_executor_job(
+                    self.api.tlx_detail, self.device_id
+                )
+                self.data = tlx_info["data"]
+            elif self.device_type == "storage":
+                storage_info_detail = await self.hass.async_add_executor_job(
+                    self.api.storage_params, self.device_id
+                )
+                storage_energy_overview = await self.hass.async_add_executor_job(
+                    self.api.storage_energy_overview, self.plant_id, self.device_id
+                )
+                self.data = {
+                    **storage_info_detail["storageDetailBean"],
+                    **storage_energy_overview,
+                }
+            elif self.device_type == "mix":
+                mix_info = await self.hass.async_add_executor_job(
+                    self.api.mix_info, self.device_id
+                )
+                mix_totals = await self.hass.async_add_executor_job(
+                    self.api.mix_totals, self.device_id, self.plant_id
+                )
+                mix_system_status = await self.hass.async_add_executor_job(
+                    self.api.mix_system_status, self.device_id, self.plant_id
+                )
+                mix_detail = await self.hass.async_add_executor_job(
+                    self.api.mix_detail, self.device_id, self.plant_id
+                )
+
+                # Get the chart data and work out the time of the last entry
+                mix_chart_entries = mix_detail["chartData"]
+                sorted_keys = sorted(mix_chart_entries)
+
+                # Create datetime from the latest entry
+                date_now = dt_util.now().date()
+                last_updated_time = dt_util.parse_time(str(sorted_keys[-1]))
+                mix_detail["lastdataupdate"] = datetime.datetime.combine(
+                    date_now, last_updated_time, dt_util.get_default_time_zone()
+                )
+
+                # Dashboard data for mix system
+                dashboard_data = await self.hass.async_add_executor_job(
+                    self.api.dashboard_data, self.plant_id
+                )
+                dashboard_values_for_mix = {
+                    "etouser_combined": float(
+                        dashboard_data["etouser"].replace("kWh", "")
+                    )
+                }
+                self.data = {
+                    **mix_info,
+                    **mix_totals,
+                    **mix_system_status,
+                    **mix_detail,
+                    **dashboard_values_for_mix,
+                }
+            _LOGGER.debug(
+                "Finished updating data for %s (%s)",
+                self.device_id,
+                self.device_type,
+            )
+        except json.decoder.JSONDecodeError:
+            _LOGGER.error("Unable to fetch data from Growatt server")
```

Comment:
If we can't fetch data for whatever reason, we should raise `UpdateFailed`. If this happens during startup (using `async_config_entry_first_refresh`) the config entry will retry later. If this happens during runtime, the attached entities will go unavailable (which is correct as we don't get any info for it)